fairseq/modules/moe/moe_layer.py

```diff
@@ -153,14 +161,23 @@ def forward(self, *input: Tensor, input_padding_mask=None, **kwargs: Any) -> Ten
                 padded_input_padding_mask[:reshaped_input_shape[0]] = False
             reshaped_input_padding_mask = padded_input_padding_mask
 
-        l_aux, combine_weights, dispatch_mask, self.metadata = self.gate(reshaped_input, reshaped_input_padding_mask)
+        if has_tutel:
+            l_aux, self.metadata, C, E, indices_, locations_, gates_ = self.gate(reshaped_input, reshaped_input_padding_mask)
+            S, M = reshaped_input.size(0), reshaped_input.size(1)
+
+            if not hasattr(self, '_tutel_dispatcher'):
+                self._tutel_dispatcher = tutel_moe.fast_dispatcher(E, C, M, dispatch_dtype=reshaped_input.dtype)
```

Comment:
Since we are initializing the `fast_dispatcher` with `E`, `C`, `M` values from the first forward, this will create issue when the capacity `C` changes (either when number of tokens are not equal to max_tokens or during eval). In the `.update` call, we should also update the `self.capacity` in the  `fast_dispatcher` for every forward.

cc @ngoyal2707

Reply:
Thanks for this requirement. The JIT kernels have to handle dynamic size C to satisfy that, while current version only consider S (batch * tokens) as dynamic values. @ngoyal2707 Does your check also target to the same case?

Reply:
@vedanuj I just append a second commit for this PR to allow `capacity` to be reset when calling `_tutel_dispatcher.update(..)`.